core/src/main/java/io/grpc/ClientCallImpl.java

```diff
@@ -0,0 +1,347 @@
+    this.scheduler = checkNotNull(scheduler, "No scheduler provided");
+    unaryRequest =
+        method.getType() == MethodType.UNARY || method.getType() == MethodType.SERVER_STREAMING;
+  }
+
+  /**
+   * Provider of {@link ClientTransport}s.
+   */
+  interface ClientTransportProvider {
+    /**
+     * @return a client transport, or null if no more transports can be created.
+     */
+    ClientTransport get();
+  }
+
+  private enum State {INIT, STARTED, HALF_CLOSED, CANCELED}
+
+  synchronized ClientCallImpl<ReqT, RespT> setUserAgent(@Nullable String userAgent) {
+    checkState(state == State.INIT, "Already started");
+    this.userAgent = userAgent;
+    return this;
+  }
+
+  @Override
+  public synchronized void start(ClientCall.Listener<RespT> callListener, Headers headers) {
+    checkState(state == State.INIT, "Already started");
+    state = State.STARTED;
+
+    // We need to keep a copy of the call listener because the stream may not be set up in time.
+    // This would happen if the transport is not ready, but the call deadline is reached, and we
+    // need to notify the callListener.
+    this.callListener = checkNotNull(callListener, "No call listener provided");
+
+    ClientStreamListener streamListener = new ClientStreamListenerImpl();
+    Status errorStatus = prepareHeaders(headers, callOptions, userAgent);
+    if (errorStatus != null) {
+      closeCallPrematurely(streamListener, errorStatus);
+      return;
+    }
+
+    boolean started = startInternal(streamListener, headers);
+
+    // Start the deadline timer after stream creation because it will close the stream
+    if (started && headers.containsKey(TIMEOUT_KEY)) {
+      deadlineCancellationFuture = startDeadlineTimer(headers.get(TIMEOUT_KEY));
+    }
+  }
+
+  /**
+   * Modifies headers to send for this call.
+   *
+   * @return null on success, or a failure status
+   */
+  private static Status prepareHeaders(
+      Headers headers, CallOptions callOptions, @Nullable String userAgent) {
+    // Fill out timeout on the headers
+    headers.removeAll(TIMEOUT_KEY);
+    // Convert the deadline to timeout. Timeout is more favorable than deadline on the wire
+    // because timeout tolerates the clock difference between machines.
+    Long deadlineNanoTime = callOptions.getDeadlineNanoTime();
+    long timeoutMicros = 0;
+    if (deadlineNanoTime != null) {
+      timeoutMicros = TimeUnit.NANOSECONDS.toMicros(deadlineNanoTime - System.nanoTime());
+      if (timeoutMicros <= 0) {
+        return Status.DEADLINE_EXCEEDED;
+      }
+      headers.put(TIMEOUT_KEY, timeoutMicros);
+    }
+
+    // Fill out the User-Agent header.
+    headers.removeAll(HttpUtil.USER_AGENT_KEY);
+    if (userAgent != null) {
+      headers.put(HttpUtil.USER_AGENT_KEY, userAgent);
+    }
+    return null;
+  }
+
+  private boolean startInternal(final ClientStreamListener streamListener, final Headers headers) {
```

Comment:
So startInternal wasn't part of the necessary. :-/ The point of doing the move separately was so that we wouldn't have to manually discover unrelated changes like this and holding a lock during transport.newStream.

I guess this will need a full review.